### ISSUE TITLE
Compare the missing value with the top value even after the hit queue is full

### ISSUE THREAD

**Issue by bugmakerrrrrr** [apache/lucene#13643]
Today, when the hit queue is full, we only compare the missing value with the bottom to decide whether it is competitive or not, but this can lead to misjudgment. For example, if the missing value is 2, the after value is 3, and when the hit queue is full, the bottom value is 6. In current implementation, the missing value is competitive, but obviously, it is out the range of match.

https://github.com/apache/lucene/blob/f0558ae26392e668e85e63710b9dada5a79a480f/lucene/core/src/java/org/apache/lucene/search/comparators/NumericComparator.java#L345-L354

**Comment by gsmiller** [comment]
Thanks for finding this (and for adding a test case)! Would it be possible to write your test such that it fails without your proposed change? As it's currently written, I believe it will pass even without your change. That would make the testing a little stronger. WDYT?

**Comment by bugmakerrrrrr** [comment]
> As it's currently written, I believe it will pass even without your change. 

@gsmiller Yes, you are right. The method `assertNonCompetitiveHitsAreSkipped` cannot make a right assertion, including the existing tests have the same problem. I wrote the test to compare the number of total hits  before and after this change(14806+ vs. 513+). I will try to write a new test.

**Comment by gsmiller** [comment]
Thanks @bugmakerrrrrr. I was thinking about this change a bit more and wondering how you came across the issue. I definitely agree there's an opportunity here to tighten up the logic as you propose, but I wonder if it's possible for this method to ever be invoked with a document that's "before" an established "after" value. Did you actually hit this issue in the wild? I don't know the "wiring" of the field sort + "after" logic that well (and haven't had time to look through it in more detail in response to your PR), but I'm curious if there's actually a reachable code-path to hit this condition. Not opposed to adding this additional logic, just curious.

**Comment by bugmakerrrrrr** [comment]
> I wonder if it's possible for this method to ever be invoked with a document that's "before" an established "after" value. 

This method is related to missing values, if there are some docs missing target field, this method will always be invoked to check whether missing value is competitive or not. 

> Did you actually hit this issue in the wild?

No. But I believe that this can be hit. For example, in OpenSearch, when we sort results on `long` field as asc order, we set `missing` value as `_first` to filter missing doc(which turns into `Long.MIN_VALUE` in Lucene) and `search_after` as 3, in this case, the scenario is same as the test that I posted.
